File: 3734.lexicographically-smallest-palindromic-permutation-greater-than-target.go
#
# @lc app=leetcode id=3734 lang=golang
#
# [3734] Lexicographically Smallest Palindromic Permutation Greater Than Target
#

# @lc code=start
func lexPalindromicPermutation(s string, target string) string {
	n := len(s)
	freq := [26]int{}
	for i := 0; i < n; i++ {
		freq[int(s[i]) - int('a')]++
	}
	odds := 0
	for i := 0; i < 26; i++ {
		if freq[i] % 2 == 1 {
			odds++
		}
	}
	if odds != n % 2 {
		return ""
	}
	halfLen := n / 2
	isOdd := n % 2 == 1
	m := halfLen + (n % 2)
	first := make([]byte, m)
	cfreq := [26]int{}
	copy(cfreq[:], freq[:])
	alreadyGreater := false
	for j := 0; j < m; j++ {
		isPair := j < halfLen
		need := 2
		if !isPair {
			need = 1
		}
		found := false
		for chb := byte('a'); chb <= byte('z'); chb++ {
			i := int(chb - 'a')
			if cfreq[i] < need {
				continue
			}
			cfreq[i] -= need
			// feasible?
			assignedPairs := j + 1
			if !isPair {
				assignedPairs = halfLen
			}
			remPairs := halfLen - assignedPairs
			remMiddle := 0
			if isOdd && j+1 < m {
				remMiddle = 1
			}
			expChars := 2 * remPairs + remMiddle
			sumRem := 0
			oddRem := 0
			for k := 0; k < 26; k++ {
				sumRem += cfreq[k]
				if cfreq[k] % 2 == 1 {
					oddRem++
				}
			}
			feasible := sumRem == expChars && oddRem == remMiddle
			if !feasible {
				cfreq[i] += need
				continue
			}
			// condition
			thisGT := alreadyGreater || (chb > target[j])
			if thisGT {
				found = true
				first[j] = chb
				alreadyGreater = alreadyGreater || (chb > target[j])
				break
			} else if chb < target[j] {
				cfreq[i] += need
				continue
			} else {
				// == case, check largest
				pMax := make([]byte, n)
				for p := 0; p < j; p++ {
					pMax[p] = first[p]
					pMax[n-1-p] = first[p]
				}
				pMax[j] = chb
				if j < halfLen {
					pMax[n-1-j] = chb
				}
				var tfreq [26]int
				copy(tfreq[:], cfreq[:])
				simSuccess := true
				for p := j + 1; p < m; p++ {
					pPair := p < halfLen
					pNeed := 2
					if !pPair {
						pNeed = 1
					}
					fMax := false
					for cc := byte('z'); cc >= byte('a'); cc-- {
						ii := int(cc - 'a')
						if tfreq[ii] >= pNeed {
							pMax[p] = cc
							if p < halfLen {
								pMax[n-1-p] = cc
							}
							tfreq[ii] -= pNeed
							fMax = true
							break
						}
					}
					if !fMax {
						simSuccess = false
						break
					}
				}
				if simSuccess && string(pMax) > target {
					found = true
					first[j] = chb
					alreadyGreater = alreadyGreater || (chb > target[j])
					break
				}
				cfreq[i] += need
			}
		}
		if !found {
			return ""
		}
	}
	res := make([]byte, n)
	for i := 0; i < halfLen; i++ {
		res[i] = first[i]
		res[n-1-i] = first[i]
	}
	if isOdd {
		res[halfLen] = first[halfLen]
	}
	return string(res)
}
# @lc code=end

File: 3734.lexicographically-smallest-palindromic-permutation-greater-than-target.java
#
# @lc app=leetcode id=3734 lang=java
#
# [3734] Lexicographically Smallest Palindromic Permutation Greater Than Target
#

# @lc code=start
class Solution {
    public String lexPalindromicPermutation(String s, String target) {
        int n = s.length();
        int[] total = new int[26];
        for (char ch : s.toCharArray()) {
            total[ch - 'a']++;
        }
        int leftEnd = (n - 1) / 2;
        for (int k = n - 1; k >= 0; k--) {
            char[] res = new char[n];
            boolean validPrefix = true;
            for (int j = 0; j < k; j++) {
                char ch = target.charAt(j);
                int mir = n - 1 - j;
                if (res[j] != 0 && res[j] != ch) {
                    validPrefix = false;
                    break;
                }
                res[j] = ch;
                if (res[mir] != 0 && res[mir] != ch) {
                    validPrefix = false;
                    break;
                }
                res[mir] = ch;
            }
            if (!validPrefix) continue;
            int[] cnt = new int[26];
            System.arraycopy(total, 0, cnt, 0, 26);
            boolean feasible = true;
            for (int i = 0; i < n; i++) {
                if (res[i] != 0) {
                    int idx = res[i] - 'a';
                    cnt[idx]--;
                    if (cnt[idx] < 0) {
                        feasible = false;
                        break;
                    }
                }
            }
            if (!feasible) continue;
            int mirK = n - 1 - k;
            char tgtK = target.charAt(k);
            char chosenC = 0;
            boolean mirrorAlreadySet = (mirK < k);
            int deduct = 1;
            if (mirK > k) deduct = 2;
            if (mirrorAlreadySet) {
                char forced = res[mirK];
                int idx = forced - 'a';
                if (forced > tgtK && cnt[idx] >= 1) {
                    chosenC = forced;
                }
            } else {
                for (char c = (char)(tgtK + 1); c <= 'z'; c++) {
                    int idx = c - 'a';
                    if (cnt[idx] >= deduct) {
                        chosenC = c;
                        break;
                    }
                }
            }
            if (chosenC == 0) continue;
            res[k] = chosenC;
            if (mirK != k) {
                res[mirK] = chosenC;
            }
            cnt[chosenC - 'a'] -= deduct;
            boolean canFill = true;
            for (int i = k + 1; i <= leftEnd; i++) {
                if (res[i] != 0) continue;
                int mirI = n - 1 - i;
                int need = (mirI == i ? 1 : 2);
                char d = 0;
                for (char cc = 'a'; cc <= 'z'; cc++) {
                    int idx = cc - 'a';
                    if (cnt[idx] >= need) {
                        d = cc;
                        break;
                    }
                }
                if (d == 0) {
                    canFill = false;
                    break;
                }
                res[i] = d;
                res[mirI] = d;
                cnt[d - 'a'] -= need;
            }
            if (!canFill) continue;
            feasible = true;
            for (int i = 0; i < n; i++) {
                if (res[i] == 0) {
                    feasible = false;
                    break;
                }
            }
            if (!feasible) continue;
            for (int i = 0; i < 26; i++) {
                if (cnt[i] != 0) {
                    feasible = false;
                    break;
                }
            }
            if (feasible) {
                return new String(res);
            }
        }
        return "";
    }
}
# @lc code=end